Source/Core/VideoBackends/Vulkan/main.cpp

```diff
@@ -89,11 +98,51 @@ static bool ShouldEnableDebugReports(bool enable_validation_layers)
   return enable_validation_layers || IsHostGPULoggingEnabled();
 }
 
+// Helpers to manage the connection to the X server.
+bool VideoBackend::OpenDisplayConnection()
+{
+#if defined(VK_USE_PLATFORM_XCB_KHR)
+  // We could probably use a xcb_connection_t here, however since we get passed an XLib window
+  // handle, not a xcb_window_t handle, to keep things consistent we will use an XLib connection.
+  m_display_handle = XOpenDisplay(nullptr);
+  if (!m_display_handle)
+  {
+    PanicAlert("Failed to open X display.");
+    return false;
+  }
+#endif
+
+  return true;
+}
+
+void VideoBackend::CloseDisplayConnection()
+{
+#if defined(VK_USE_PLATFORM_XCB_KHR)
+  if (m_display_handle)
+  {
+    XCloseDisplay(reinterpret_cast<Display*>(m_display_handle));
+    m_display_handle = nullptr;
+  }
+#endif
+}
+
 bool VideoBackend::Initialize(void* window_handle)
 {
+  // If we are running headless, we don't need the surface extensions or a display connection.
+  bool enable_surface = window_handle != nullptr;
+
+  // Ensure we have a connection to the display server (if any) before continuing.
+  if (enable_surface && !OpenDisplayConnection())
+    return false;
+
+  // Dynamically load vulkan-1.dll/libvulkan.so.
   if (!LoadVulkanLibrary())
   {
-    PanicAlert("Failed to load Vulkan library.");
+    PanicAlert("Failed to load Vulkan library.\n\n"
+               "This may be due to the loader not being installed on your system.\n\n"
+               "The loader is usually provided with your display driver on Windows, "
+               "or in a package named 'libvulkan1' or 'vulkan-icd-loader' on Linux.");
+    CloseDisplayConnection();
```

Comment:
Assuming video backends get freed if initialization fails. You could possibly make `m_display_handle` a `std::unique_ptr` and just supply it with a deleter containing the `XCloseDisplay` call in `CloseDisplayConnection()`, which would make it less repetitive throughout the function/safer overall.